edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/NestedConnectionTrackingTest.cs

```diff
@@ -0,0 +1,587 @@
+                }
+
+                return this.playbook[(int)eventType, phase];
+            }
+
+            void GenerateSubscriptions(int minDuration, int maxDuration, bool isModule)
+            {
+                // start from 1 as for 'Message' a different loop makes the generation
+                for (int i = 1; i < (int)SubscriptionOrMessage.Last; i++)
+                {
+                    // Modules cannot handle C2D, so skip generating those
+                    if (i == (int)SubscriptionOrMessage.C2D && isModule)
+                    {
+                        continue;
+                    }
+
+                    var pos = 0;
+                    while (pos < this.phases)
+                    {
+                        var offLen = Math.Min(rnd.Next(minDuration, maxDuration), this.phases - pos);
+
+                        for (var n = 0; n < offLen; n++)
+                        {
+                            this.playbook[i, pos++] = false;
+                        }
+
+                        if (pos < this.phases)
+                        {
+                            var onLen = Math.Min(rnd.Next(minDuration, maxDuration), phases - pos);
+
+                            for (var n = 0; n < onLen; n++)
+                            {
+                                this.playbook[i, pos++] = true;
+                            }
+                        }
+                    }
+                }
+            }
+
+            void GenerateMessages(int maxDuration)
+            {
+                var pos = 0;
+                while (pos < this.phases)
+                {
+                    var nextPos = Math.Min(rnd.Next(maxDuration), this.phases - pos);
+                    for (var n = 0; n < nextPos; n++)
+                    {
+                        this.playbook[0, pos++] = false;
+                    }
+
+                    if (pos < this.phases)
+                    {
+                        this.playbook[0, pos++] = true;
+                    }
+                }
+            }
+        }
+
+        List<TestClient> GenerateClients(int clientCount, double moduleRatio, double directRatio)
+        {
+            var rnd = new Random(921752);
+            var result = new List<TestClient>();
+
+            for (var i = 0; i < clientCount; i++)
+            {
+                bool isModule = moduleRatio > rnd.NextDouble();
+                bool isDirect = directRatio > rnd.NextDouble();
+
+                var identity = isModule ? new ModuleIdentity(iotHubName, "device_" + i.ToString(), "module_" + i.ToString()) as IIdentity
+                                        : new DeviceIdentity(iotHubName, "device_" + i.ToString()) as IIdentity;
+
+                var newClient = new TestClient(isDirect, identity);
+
+                result.Add(newClient);
+            }
+
+            return result;
+        }
+
+        [Fact]
+        public async Task ConnectionsAndSubscriptionsAreBeingTracked()
```

Comment:
it would be great to describe the scenario in method comments. It is very hard to follow such a huge test.